throttle.go

```diff
@@ -0,0 +1,55 @@
+package taskgroup
+
+// A Throttle rate-limits the number of concurrent goroutines that can execute
+// in parallel to some fixed number.  A zero Throttle is ready for use, but
+// imposes no limit on parallel execution. See [Throttle.Enter] for use.
+type Throttle struct {
+	adm chan struct{}
+}
+
+// NewThrottle constructs a [Throttle] with a capacity of n goroutines.
+// If n ≤ 0, the resulting Throttle imposes no limit.
+func NewThrottle(n int) Throttle {
+	if n <= 0 {
+		return Throttle{}
+	}
+	return Throttle{adm: make(chan struct{}, n)}
+}
+
+// Enter blocks until a slot is available in t, then returns a [Leaver] that
+// the caller must execute to return the slot when it is no longer in use.
+func (t Throttle) Enter() Leaver {
+	if t.adm == nil {
+		return func() {}
+	}
+	t.adm <- struct{}{}
+	var done bool
```

Comment:
if you made this an atomic.Bool and CAS it, you could remove the caveat on when it's safe to call Leaver, for low runtime cost (I think? One wacky cache-coherent memory cycle vs. a core-local incoherent one)

Reply:
Yeah, I considered doing that, and the fact that you thought it was worth mentioning suggests it's probably better to just do it. In the common case it will only be run once anyway.